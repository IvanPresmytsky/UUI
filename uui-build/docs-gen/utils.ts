import { TTypeProp } from './types';
import fs from 'fs';

export function sortProps(propsArr: TTypeProp[]): TTypeProp[] {
    if (!propsArr) {
        return undefined;
    }
    const arr = [...propsArr];
    arr.sort((p1, p2) => {
        return compareStr(p1.name, p2.name)
            || compareStr(`${p1.inheritedFrom.module}/${p1.inheritedFrom.name}`, `${p2.inheritedFrom.module}/${p2.inheritedFrom.name}`)
            || compareStr(p1.optional, p2.optional);
    });
    return arr;
}

function compareStr(s1?: string | boolean, s2?: string | boolean) {
    return String(s1).localeCompare(String(s2));
}

export function saveJsonToFile(fullPath: string, jsonToSave: object) {
    if (fs.existsSync(fullPath)) {
        fs.rmSync(fullPath, { force: true });
    }
    const content = JSON.stringify(jsonToSave, undefined, 1);
    fs.writeFileSync(fullPath, content);
}

export function isExternalFile(filePath: string) {
    return filePath.indexOf('node_modules') !== -1 && filePath.indexOf('@epam/') === -1;
}
