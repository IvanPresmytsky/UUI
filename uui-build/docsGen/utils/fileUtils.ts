import fs from 'fs';
import path from 'path';
import { uuiRoot } from '../constants';
import { INCLUDED_UUI_PACKAGES } from '../constants';

export function makeRelativeToUuiRoot(fullPath: string) {
    return path.relative(uuiRoot, fullPath).replace(/\\/g, '/');
}

export function saveContentToFile(fullPath: string, contentToSave: object | string) {
    if (fs.existsSync(fullPath)) {
        fs.rmSync(fullPath, { force: true });
    }
    const content = typeof contentToSave === 'string' ? contentToSave : JSON.stringify(contentToSave, undefined, 1);
    fs.writeFileSync(fullPath, content);
}

export function isExternalFile(filePath: string) {
    return filePath.indexOf('node_modules') !== -1 && filePath.indexOf('@epam/') === -1;
}

export function getUuiModuleNameFromPath(absolutePath: string) {
    const rel = path.relative(uuiRoot, absolutePath);
    const moduleFolderName = rel.split(path.sep)[0];
    const foundEntry = Object.entries(INCLUDED_UUI_PACKAGES).find((e) => {
        const folderName = e[1];
        return folderName === moduleFolderName;
    });
    if (foundEntry) {
        return foundEntry[0];
    }
    return rel.replace(/\\/g, '/');
}
