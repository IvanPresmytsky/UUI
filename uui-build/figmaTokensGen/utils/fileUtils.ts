import fs from 'fs';
import path from 'path';
import { IFigmaVarCollection } from '../types';
import { logger, uuiRoot, PATH } from '../constants';
import { FigmaScriptsContext } from '../context/context';

export class FileUtils {
    static writeResults(params: { result: IFigmaVarCollection, ctx: FigmaScriptsContext }) {
        const { result, ctx } = params;
        const outDir = path.resolve(uuiRoot, PATH.RESULTS_DIR);
        const extName = path.extname(PATH.FIGMA_VARS_COLLECTION_SRC);
        const sourceFileNameNoExt = path.basename(PATH.FIGMA_VARS_COLLECTION_SRC, extName);
        const resultPath = forwardSlashes(path.resolve(outDir, `${sourceFileNameNoExt}-generated${extName}`));
        const logPath = forwardSlashes(path.resolve(outDir, `${sourceFileNameNoExt}-log${extName}`));
        fs.writeFileSync(resultPath, JSON.stringify(result, undefined, 2));
        fs.writeFileSync(logPath, JSON.stringify(ctx.log.toJson(), undefined, 2));
        logger.success(`Files created:\n${resultPath}\n${logPath}`);
    }

    static readFigmaVarCollection(): IFigmaVarCollection {
        const absPath = path.resolve(uuiRoot, PATH.FIGMA_VARS_COLLECTION_SRC);
        if (fs.existsSync(absPath)) {
            const content = fs.readFileSync(absPath);
            return JSON.parse(content.toString());
        }
        throw new Error(`Figma var collection not found: ${absPath}`);
    }
}

function forwardSlashes(pathStr: string) {
    return pathStr.replace(/\\/g, '/');
}
