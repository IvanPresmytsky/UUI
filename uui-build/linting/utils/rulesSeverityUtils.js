const { isLintStaged } = require('../../utils/envUtils.js');

const SEVERITY = {
    off: 'off',
    warn: 2, // during lint staged warnings don't prevent from commit, so we make them as errors.
    error: 2,
};

function setSeverity(ruleConfig, severity) {
    if (Array.isArray(ruleConfig)) {
        ruleConfig[0] = severity;
        return ruleConfig;
    }
    if (typeof ruleConfig === 'number' || typeof ruleConfig === 'string') {
        return severity;
    }
    throw new Error('Unexpected rule config', ruleConfig);
}

function overrideSeverityInRulesMap(rulesMap) {
    return Object.keys(rulesMap).reduce((acc, ruleName) => {
        acc[ruleName] = setSeverity(rulesMap[ruleName], SEVERITY.warn);
        return acc;
    }, {});
}

module.exports = { SEVERITY, setSeverity, overrideSeverityInRulesMap };
