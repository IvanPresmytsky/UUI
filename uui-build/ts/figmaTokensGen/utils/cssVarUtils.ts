import {
    IFigmaVar, IFigmaVarCollection, TFigmaVariableAlias,
    TRgbaValue,
} from '../types/sourceTypes';
import { CONFIG } from '../config';
import { rgbaToHEXA } from './colorUtils';
import {
    IThemeVar,
    TCssVarRef,
    TFigmaThemeName,
    TFloatValue,
    TResolvedValueNorm,
    TUuiCssVarName,
    TVarType,
} from '../types/sharedTypes';

export function getCssVarFromFigmaVar(figmaVarPath: string): TUuiCssVarName {
    const tokens = figmaVarPath.split('/');
    const lastToken = tokens[tokens.length - 1];
    return `--uui-${lastToken}`;
}

export function isCssVarSupportedByUui(figmaVarPath: string) {
    const isIgnoredByConvention = figmaVarPath.indexOf('core/') !== 0;
    const cssVarName = getCssVarFromFigmaVar(figmaVarPath);
    const isIgnoredByConfig = CONFIG.tokens[cssVarName]?.supported === false;
    return !isIgnoredByConvention && !isIgnoredByConfig;
}

export function getNormalizedResolvedValueMap(
    params: {
        figmaVar: IFigmaVar,
        modes: IFigmaVarCollection['modes'],
        figmaVarById: Record<string, IFigmaVar>
    },
): IThemeVar['valueByTheme'] {
    const { figmaVar, modes, figmaVarById } = params;
    const themeIdArr = Object.keys(modes);

    return themeIdArr.reduce<IThemeVar['valueByTheme']>((acc, themeId) => {
        const themeName = modes[themeId] as TFigmaThemeName;
        const valueChain = getNormalizedResolvedValue({ figmaVar, themeId, figmaVarById, mode: 'chain' });
        const valueDirect = getNormalizedResolvedValue({ figmaVar, themeId, figmaVarById, mode: 'direct' });

        acc[themeName] = {
            valueChain,
            valueDirect,
        };
        return acc;
    }, {});
}

function getNormalizedResolvedValue(
    params: {
        figmaVar: IFigmaVar,
        figmaVarById: Record<string, IFigmaVar>,
        themeId: string,
        mode: 'chain' | 'direct'
    },
): TResolvedValueNorm {
    const { figmaVarById, mode, themeId } = params;
    if (params.mode === 'direct') {
        const item = params.figmaVar.resolvedValuesByMode[params.themeId];
        const value = getNormalizedValue({ type: params.figmaVar.type, value: item.resolvedValue });
        const alias = getNormalizedAlias((item as { aliasName: string })?.aliasName);
        return {
            alias,
            value,
        };
    }
    if (params.mode === 'chain') {
        const item = params.figmaVar.valuesByMode[params.themeId];
        if ((item as TFigmaVariableAlias).type) {
            const { id } = item as TFigmaVariableAlias;
            const nextVar = params.figmaVarById[id];
            const nextVarAlias = getNormalizedAlias(nextVar.name);

            const { value, alias } = getNormalizedResolvedValue({ figmaVar: nextVar, figmaVarById, mode, themeId });
            return {
                value,
                alias: nextVarAlias.concat(alias),
            };
        } else {
            return {
                alias: [],
                value: getNormalizedValue({ type: params.figmaVar.type, value: item }),
            };
        }
    }
    throw new Error('Unknown mode');
}

// aliasName is a path
function getNormalizedAlias(aliasName: string | undefined): TCssVarRef[] {
    if (aliasName) {
        const supported = isCssVarSupportedByUui(aliasName);
        const cssVar = getCssVarFromFigmaVar(aliasName);
        return [{
            id: aliasName,
            cssVar,
            supported,
        }];
    }
    return [];
}

function getNormalizedValue(params: { type: TVarType, value: unknown }) {
    switch (params.type) {
        case TVarType.COLOR: {
            return rgbaToHEXA(params.value as TRgbaValue);
        }
        case TVarType.FLOAT: {
            return params.value as TFloatValue;
        }
    }
}
