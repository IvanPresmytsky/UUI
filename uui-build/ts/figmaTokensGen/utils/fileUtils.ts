import fs from 'fs';
import path from 'path';
import { IFigmaVarCollection } from '../types/sourceTypes';
import { PATH } from '../constants';
import { FigmaScriptsContext } from '../context/context';
import { logger, uuiRoot } from '../../jsBridge';
import { IUuiTokensCollection } from '../types/sharedTypes';

export class FileUtils {
    static writeResults(
        params: {
            newFigmaVarCollection: IFigmaVarCollection,
            uuiTokensCollection: IUuiTokensCollection,
            ctx: FigmaScriptsContext
        },
    ) {
        const { newFigmaVarCollection, uuiTokensCollection } = params;
        const outDir = path.resolve(uuiRoot, PATH.RESULTS_DIR);
        const extName = path.extname(PATH.FIGMA_VARS_COLLECTION_SRC);
        const sourceFileNameNoExt = path.basename(PATH.FIGMA_VARS_COLLECTION_SRC, extName);
        const resultPath = forwardSlashes(path.resolve(outDir, `${sourceFileNameNoExt}Output${extName}`));
        const uuiTokensCollectionPath = forwardSlashes(path.resolve(outDir, `${sourceFileNameNoExt}Tokens${extName}`));
        fs.writeFileSync(resultPath, JSON.stringify(newFigmaVarCollection, undefined, 2));
        fs.writeFileSync(uuiTokensCollectionPath, JSON.stringify(uuiTokensCollection, undefined, 2));
        logger.success(`Files created:\n${resultPath}\n${uuiTokensCollectionPath}`);
    }

    static readFigmaVarCollection(): IFigmaVarCollection {
        const absPath = path.resolve(uuiRoot, PATH.FIGMA_VARS_COLLECTION_SRC);
        if (fs.existsSync(absPath)) {
            const content = fs.readFileSync(absPath);
            return JSON.parse(content.toString());
        }
        throw new Error(`Figma var collection not found: ${absPath}`);
    }
}

function forwardSlashes(pathStr: string) {
    return pathStr.replace(/\\/g, '/');
}
