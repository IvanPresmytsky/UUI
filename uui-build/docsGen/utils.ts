import { TTypeProp, typeRefToUniqueString } from './types';
import fs from 'fs';
import path from 'path';
// @ts-ignore
import { uuiRoot } from '../utils/constants';

export function makeRelativeToUuiRoot(fullPath: string) {
    return path.relative(uuiRoot, fullPath).replace(/\\/g, '/');
}

function propsComparator(p1: TTypeProp, p2: TTypeProp) {
    function compareStr(s1: string, s2: string) {
        return String(s1).localeCompare(String(s2));
    }
    return compareStr(p1.name, p2.name)
        || compareStr(`${p1.from?.module}/${p1.from?.typeName.name}`, `${p2.from?.module}/${p2.from?.typeName.name}`)
        || compareStr(String(p1.required), String(p2.required));
}

export function sortProps(propsArr: TTypeProp[]): TTypeProp[] | undefined {
    if (propsArr) {
        return [...propsArr].sort(propsComparator);
    }
}

export function saveContentToFile(fullPath: string, contentToSave: object | string) {
    if (fs.existsSync(fullPath)) {
        fs.rmSync(fullPath, { force: true });
    }
    const content = typeof contentToSave === 'string' ? contentToSave : JSON.stringify(contentToSave, undefined, 1);
    fs.writeFileSync(fullPath, content);
}

export function isExternalFile(filePath: string) {
    return filePath.indexOf('node_modules') !== -1 && filePath.indexOf('@epam/') === -1;
}

export class SimpleIdGen {
    private _id = 0;

    getNextId = () => {
        return String(++this._id);
    };
}

export class PropsSet {
    private _propsMap = new Map<string, TTypeProp>();

    add(p: TTypeProp) {
        const id = PropsSet.buildId(p);
        this._propsMap.set(id, p);
    }

    addAll(pa: TTypeProp[]) {
        pa.forEach((p) => {
            const id = PropsSet.buildId(p);
            this._propsMap.set(id, p);
        });
    }

    has(p: TTypeProp): boolean {
        const id = PropsSet.buildId(p);
        return this._propsMap.has(id);
    }

    toArray(): TTypeProp[] {
        return [...this._propsMap.values()];
    }

    static concatAndSort(psa: PropsSet[]): TTypeProp[] {
        const tempPs = psa.reduce<PropsSet>((acc, ps) => {
            acc.addAll(ps.toArray());
            return acc;
        }, new PropsSet());
        return sortProps(tempPs.toArray());
    }

    static fromArray(pa: TTypeProp[]): PropsSet {
        const ps = new PropsSet();
        pa.forEach((p) => {
            ps.add(p);
        });
        return ps;
    }

    static buildId(p: TTypeProp) {
        // we don't use uniqueId property because we need to eliminate duplicates which come from same type (it's possible in union types)
        return `${typeRefToUniqueString(p.from)}:${p.name}:${p.typeValue.raw.replace(/[\n\s]/g, '')}`;
    }
}
