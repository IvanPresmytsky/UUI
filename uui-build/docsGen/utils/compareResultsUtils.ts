import { TDocGenStatsResult } from '../types/types';
import { OUTPUT_FILE_COMPARISON_REPORT_MD, uuiRoot } from '../constants';
import { saveContentToFile } from './fileUtils';
import fs from 'fs';
import path from 'path';

type TCompareStatsResult = {
    propsNoComment: {
        prev: number | undefined,
        next: number,
        isIncreased: boolean,
        newWithNoComment: string[],
    }
    typesNoComment: {
        prev: number | undefined,
        next: number,
        isIncreased: boolean,
        newWithNoComment: string[],
    }
};

export function compareToPrevStats(nextStats: TDocGenStatsResult) {
    removeComparisonReportMd();
    const relPath = getPrevStatsFromCliArg();
    if (relPath) {
        const prevStats = getPrevStatsFromFile(relPath);
        const comparisonResult = compareStats({ prevStats, nextStats });
        const md = formatResultsToMd(comparisonResult);
        saveContentToFile(OUTPUT_FILE_COMPARISON_REPORT_MD, md);
    }
}

function getPrevStatsFromCliArg() {
    const arg = process.argv[2];
    if (arg) {
        const [name, relPath] = arg.split('=');
        if (name === '--prev-stats') {
            return relPath;
        }
        throw new Error(`Unknown argument: ${arg}`);
    }
}

function getPrevStatsFromFile(relPath: string): TDocGenStatsResult {
    const fullPath = path.resolve(uuiRoot, relPath);
    if (fs.existsSync(fullPath)) {
        return JSON.parse(fs.readFileSync(fullPath).toString());
    } else {
        console.warn(`Prev stats not found. The file doesn't exist: ${fullPath}`);
        return;
    }
}

function compareStats(params: { prevStats: TDocGenStatsResult | undefined, nextStats: TDocGenStatsResult }): TCompareStatsResult {
    const { prevStats, nextStats } = params;

    if (prevStats) {
        const propsNoComment = {
            prev: prevStats.missingPropComment.totals.amountProps,
            next: nextStats.missingPropComment.totals.amountProps,
        };
        const typesNoComment = {
            prev: prevStats.missingTypeComment.totals.amountTypes,
            next: nextStats.missingTypeComment.totals.amountTypes,
        };
        return {
            propsNoComment: {
                ...propsNoComment,
                isIncreased: propsNoComment.next > propsNoComment.prev,
                newWithNoComment: getNewPropsWithNoComments(params),
            },
            typesNoComment: {
                ...typesNoComment,
                isIncreased: typesNoComment.next > typesNoComment.prev,
                newWithNoComment: getNewTypesWithNoComments(params),
            },
        };
    } else {
        return {
            propsNoComment: {
                prev: undefined,
                next: nextStats.missingPropComment.totals.amountProps,
                isIncreased: false,
                newWithNoComment: [],
            },
            typesNoComment: {
                prev: undefined,
                next: nextStats.missingTypeComment.totals.amountTypes,
                isIncreased: false,
                newWithNoComment: [],
            },
        };
    }
}

function getNewTypesWithNoComments(params: { prevStats: TDocGenStatsResult, nextStats: TDocGenStatsResult }): string[] {
    const res: string[] = [];
    const { prevStats, nextStats } = params;
    const prevSet = new Set(prevStats.missingTypeComment.value);
    nextStats.missingTypeComment.value.forEach((typeRef) => {
        if (!prevSet.has(typeRef)) {
            res.push(typeRef);
        }
    });
    return res;
}

function getNewPropsWithNoComments(params: { prevStats: TDocGenStatsResult, nextStats: TDocGenStatsResult }): string[] {
    const res: string[] = [];
    const { prevStats, nextStats } = params;
    const prevSet = new Set(normMissingPropComments(prevStats));
    const nextArr = normMissingPropComments(nextStats);
    nextArr.forEach((typeRef) => {
        if (!prevSet.has(typeRef)) {
            res.push(typeRef);
        }
    });
    return res;

    function normMissingPropComments(stats: TDocGenStatsResult) {
        return stats.missingPropComment.value.reduce<string[]>((acc, { typeRef, value }) => {
            value.forEach((propName) => {
                acc.push(`${typeRef}/${propName}`);
            });
            return acc;
        }, []);
    }
}

function formatResultsToMd(result: TCompareStatsResult) {
    const isErr = result.typesNoComment.isIncreased || result.propsNoComment.isIncreased;
    const typesAmount = formatAmountWithDiff(result.typesNoComment.prev, result.typesNoComment.next);
    const propsAmount = formatAmountWithDiff(result.propsNoComment.prev, result.propsNoComment.next);
    const mdGeneratedBy = 'Generated by: generate-components-api';
    const ciStatus = `CI Status: ${isErr ? 'error' : 'ok'} \n`;
    const SUMMARY_TITLE = 'New missing comments';
    const SUMMARY_DESC = 'NOTE: Comments in the entities below are missed in the report generated from merge commit, but they aren\'t missed in the target branch report.';
    let detailsMd = '';
    if (isErr) {
        detailsMd = [
            '<details>',
            `<summary>${SUMMARY_TITLE}</summary>`,
            '<br>',
            SUMMARY_DESC,
            '',
            '```',
            newWithNoCommentsToMdList('Types', result.typesNoComment.newWithNoComment),
            newWithNoCommentsToMdList('Props', result.propsNoComment.newWithNoComment),
            '```',
            '</details>',
        ].join('\n');
    }
    return [
        mdGeneratedBy,
        ciStatus,
        '**Types/Props without comments**<br>',
        '| Entity  | Amount                    |',
        '|:-------:|:-------------------------:|',
        `|  Types  |       ${typesAmount}      |`,
        `|  Props  |       ${propsAmount}      |`,
        detailsMd,
    ].join('\n');

    function newWithNoCommentsToMdList(label: string, items: string[]) {
        if (items.length > 0) {
            const mdList = items.map((line) => `- ${line}`).join('\n');
            return `${label}:\n${mdList}`;
        }
        return '<empty>';
    }
    function formatAmountWithDiff(prev: number | undefined, next: number) {
        let mdTypesAmountChange = '(no baseline :warning:)';
        if (prev !== undefined) {
            const typesAmountDiff = next - prev;
            let typesAmountSign = '';
            let emodji = ':ok:';
            if (typesAmountDiff < 0) {
                typesAmountSign = '-';
            } else if (typesAmountDiff > 0) {
                typesAmountSign = '+';
                emodji = ':no_entry:';
            }
            mdTypesAmountChange = `(${typesAmountSign}${typesAmountDiff}) ${emodji}`;
        }
        return `${next} ${mdTypesAmountChange}`;
    }
}

function removeComparisonReportMd() {
    if (fs.existsSync(OUTPUT_FILE_COMPARISON_REPORT_MD)) {
        fs.rmSync(OUTPUT_FILE_COMPARISON_REPORT_MD, { force: true });
    }
}
