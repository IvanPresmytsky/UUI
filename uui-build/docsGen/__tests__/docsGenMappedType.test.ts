import { generateDocs } from './utils/test-utils';

describe('docsGen:mappedType', () => {
    test('should convert mapped type', () => {
        const input = `
            export type TLocal = {
                [key: number]: string;
            }
        `;
        expect(generateDocs(input)).toMatchSnapshot();
    });
});
