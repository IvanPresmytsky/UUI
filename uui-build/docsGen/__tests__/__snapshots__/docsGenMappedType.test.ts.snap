// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`docsGen:mappedType should convert mapped type 1`] = `
{
 "allTypes": {
  "test/test.tsx:TLocal": {
   "summary": {
    "module": "test/test.tsx",
    "typeName": {
     "name": "TLocal",
     "nameFull": "TLocal"
    },
    "src": "test/test.tsx",
    "exported": true
   },
   "details": {
    "kind": 264,
    "typeValue": {
     "raw": "TLocal",
     "print": [
      "type TLocal = {",
      "    [key: number]: string;",
      "};"
     ]
    },
    "props": [
     {
      "uid": 1,
      "name": "[key: number]",
      "typeValue": {
       "raw": "string"
      },
      "required": true
     }
    ],
    "propsFromUnion": false
   }
  }
 }
}
`;
